### PR TITLE
Cookie Management

### LINKED ISSUE #6: Cookie Management
-  [ ] `Set-Cookie` parsing (https://tools.ietf.org/html/rfc6265#section-5.2)
-  [ ] utility function to create Set-Cookie headers (https://tools.ietf.org/html/rfc6265#section-4)
-  [ ] "cookie jar" data structure for clients (follow https://tools.ietf.org/html/rfc6265#section-5.3)
-  [ ] `Cookie` header creation (follow https://tools.ietf.org/html/rfc6265#section-5.4)

I'm happy for this to use either lpeg or plain lua patterns.

### PR DESCRIPTION
**Pull request by daurnimator**
Fixes #6 
Closes #79 

I took a few pieces out of #79, but the initial version of this is based on code I wrote on a plane in January 2017. It tries to keep close to RFC 6265. The later revisions are cribbing from https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-02

Brings in two new dependencies: lua-psl and binaryheap

### CODE REVIEW

File: http/cookies.lua [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+				-- 1. Their name matches the name of the newly-created cookie.
+				local cmp_cookie = path_cookies[name]
+				if cmp_cookie.expiry_time < now then
+					self:clean()
+				elseif
+					-- 2. Their secure-only-flag is true.
+					cmp_cookie.secure
+					-- 3. Their domain domain-matches the domain of the newly-created
+					-- cookie, or vice-versa.
+					and domain_match(cookie.domain, cmp_cookie.domain)
```
Comment (daurnimator) [inline http/cookies.lua]: Move this to the outer loop.

File: http/cookies.lua [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+	return true
+end
+
+function store_methods:lookup(req_domain, req_path, req_is_http, req_is_secure)
+	assert(type(req_domain) == "string")
+	assert(type(req_path) == "string")
+	local now = self.time()
+	local list = {}
+	local n = 0
+	for domain, domain_cookies in pairs(self.domains) do
```
Comment (daurnimator) [inline http/cookies.lua]: Needs optimising: can index by each `req_domain` domain+superdomain?

**Comment by daurnimator** [comment]
Need to add libpsl to travis-ci.

**Comment by daurnimator** [comment]
Add `:next_clean()` method?

File: http/cookie.lua [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+end
+
+local canonicalise_host
+if has_psl then
+	canonicalise_host = psl.str_to_utf8lower
+else
+	canonicalise_host = function(str)
+		-- fail on non-ascii chars
+		if str:find("[^%p%w]") then
+			return nil
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply http/cookie.lua]: Sadly `libpsl` doesn't tell us why we failed; so can't do :(

File: http/cookie.lua [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+			path_cookies = domain_cookies[cookie.path]
+			if path_cookies ~= nil then
+				old_cookie = path_cookies[name]
+			end
+		end
+
+		-- If the cookie store contains a cookie with the same name,
+		-- domain, and path as the newly created cookie:
+		if old_cookie then
+			-- If the newly created cookie was received from a "non-HTTP"
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply http/cookie.lua]: Correct. e.g. if it was set with `document.cookie`.
Think of it as a flag that decides if the user can change it vs only transport

File: http/cookie.lua [hunk 5]
```diff
@@ -0,0 +1,861 @@
...
+	local domain = params.domain or "";
+
+	-- If the first character of the attribute-value string is %x2E ("."):
+	-- Let cookie-domain be the attribute-value without the leading %x2E (".") character.
+	if domain:sub(1, 1) == "." then
+		domain = domain:sub(2)
+	end
+
+	-- Convert the cookie-domain to lower case.
+	domain = domain:lower()
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply http/cookie.lua]: I'm not sure.... the specs seem quiet on the issue?

File: http/cookie.lua [hunk 6]
```diff
@@ -0,0 +1,861 @@
...
+	-- use the fields that make a cookie unique
+	if a.domain ~= b.domain then
+		return a.domain < b.domain
+	end
+	return a.name < b.name
+end
+
+local function cookie_match(cookie, req_domain, req_is_http, req_is_secure, req_is_safe_method, req_site_for_cookies, req_is_top_level)
+	if cookie.host_only then -- Either:
+		-- The cookie's host-only-flag is true and the canonicalized
```
[the remark this answers is not in the document]
Comment (daurnimator) [reply http/cookie.lua]: How do? It's canonicalized at the call site. (and `cookie_match` isn't public)

[the remark this answers is not in the document]
Comment (RyanSquared) [reply http/cookie.lua]: That would make sense then, but it's not obvious it's canonicalized elsewhere. Perhaps add a comment on that?

**Pull request merged**